Updates  to oslo-aero including performance

mvertens opened the pull request:
This PR does the following:
- updates oslo-aero (and in particular oslo_aero/src_cam) to be consistent with the changes that will be brought into CAM with an updated branch (feature/UpdateNorESMBranch) 
- implements significant performance updates to oslo-aero
Testing:
- regression testing was carried out using an updated test suite - this will be summarized shortly!

mvertens committed:
updates of some shadow file for new version of cam

mvertens committed:
updated radiation.F90 to be consistent with updated rrtmg

mvertens committed:
added timing

mvertens committed:
improvement of timing for oslo_aero_microphys

mvertens committed:
more performance updates

mvertens committed:
performance updates - bfb

mvertens committed:
more performane updates

gold2718 commented on src/oslo_aero_coag.F90:
```diff
@@ -131,6 +135,7 @@ subroutine initializeCoagulation(rhob,rk)
     character(8)                   :: unit
     logical                        :: history_aerosol
     logical                        :: isAlreadyOnList(gas_pcnst)
+    integer                        :: m,l
```
Single letter indices are really hard to track. Please use meaningful (short) names such as `modeind` and `tracind` or even `mind` and `tind`.

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_conc.F90:
```diff
@@ -82,26 +85,48 @@ subroutine oslo_aero_conc_calc(ncol, mmr, rho_air, CProcessModes, &
     real(r8) :: f_nbc(pcols,pver)   ! [-] mass fraction of bc in uncoated bc/oc mode
     real(r8) :: f_soana(pcols,pver) ! [-]
+    integer  :: kcomp
+    integer  :: l
```
Please do not use single letter variables. `lind`, `ind`, or `index` would be better here.

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_conc.F90:
```diff
@@ -207,18 +232,22 @@ subroutine getAerosolMask(ncol,numberConcentration, hasAerosol)
+
+    integer :: k !counter for levels
+    integer :: m !counter for modes
+    integer :: i !counter for columns
```
How about calling the variables `lev`, `mode`, and `col`?

gold2718 commented on src/oslo_aero_conc.F90:
```diff
@@ -249,8 +278,8 @@ subroutine calculateHygroscopicity(ncol, mmr, numberConcentration, rho_air, Cam,
     integer  :: kcomp !counter for modes
     integer  :: l     !counter for components
     integer  :: k     !counter for levels
+    integer  :: i     !counter for columns
```
Another opportunity to make the code easier to read.

mvertens replied:
Done.

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_conc.F90:
```diff
@@ -207,18 +232,22 @@ subroutine getAerosolMask(ncol,numberConcentration, hasAerosol)
-       end do  !levels
-    end do     !modes
+          do i=1,ncol
+             if (numberConcentration(i,k,m) .gt. smallNumber) then
```
Use modern syntax?
```suggestion
             if (numberConcentration(i,k,m) > smallNumber) then
```

mvertens replied:
Done

gold2718 commented on src/oslo_aero_conc.F90:
```diff
@@ -361,56 +377,46 @@ subroutine calculateHygroscopicity(ncol, mmr, numberConcentration, rho_air, Cam,
-
+          do i = 1,ncol
+             if (hasAerosol(i,k,kcomp)) then
+                if (VolumeCoat(i,k,kcomp) .gt. 1.e-30_r8) then
```
Modern syntax?
```suggestion
                if (VolumeCoat(i,k,kcomp) > 1.e-30_r8) then
```

mvertens replied:
Done

gold2718 commented on src/oslo_aero_conc.F90:
```diff
@@ -361,56 +377,46 @@ subroutine calculateHygroscopicity(ncol, mmr, numberConcentration, rho_air, Cam,
+                     / numberConcentration(i,k,kcomp)*(6.0_r8/pi))**athird
+
+                ! use one or the other hygroscopicity based on coating
+                if ( averageRadiusTotal(i,k) - averageRadiusCore(i,k)  .gt. coatingLimit ) then
```
Modern syntax?
```suggestion
                if ( averageRadiusTotal(i,k) - averageRadiusCore(i,k)  > coatingLimit ) then
```

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_conc.F90:
```diff
@@ -430,8 +436,9 @@ subroutine addModeHygroscopicity (ncol, hasAerosol, massConcentrationTracerInMod
-    integer                 :: k                                        !counter for levels
+    real(r8) :: massFractionInCoating
+    integer  :: k ! counter for levels
+    integer  :: i ! counter for columns
```
Please use findable index names.

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_condtend.F90:
```diff
@@ -108,6 +111,7 @@ subroutine initializeCondensation()
     real(r8) :: radmol           ![m] radius molecule
     integer  :: iDonor
     integer  :: l_donor
+    integer  :: m, l
```
Please use findable indices. For instance:
```suggestion
    integer  :: mode, trac
```

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_depos.F90:
```diff
@@ -125,11 +128,20 @@ subroutine oslo_aero_depos_init( pbuf2d )
+    do m = 0,nmodes
+       num_tracers_in_mode(m) = getNumberOfTracersInMode(m)
+    end do
+    do m = 0,nmodes
```
These two lines are not necessary.

mvertens replied:
done.

gold2718 commented on src/oslo_aero_hetfrz.F90:
```diff
@@ -166,9 +209,11 @@ subroutine hetfrz_classnuc_oslo_init(mincld_in)
     integer  :: istat
     real(r8) :: sigma_logr_aer
+    real(r8) :: sigmag_amode(3)
+    real(r8) :: m
```
Please use useful variable names. Here, `m` is a contact angle:
```suggestion
    real(r8) :: theta
```
On the other hand, it may be more efficient to eliminate it (see above).

mvertens replied:
Done. m => cos_angle

gold2718 commented on src/oslo_aero_hetfrz.F90:
```diff
@@ -100,7 +101,49 @@ module oslo_aero_hetfrz
+  ! Wang et al., 2014 fitting parameters
+  !********************************************************
+  ! freezing parameters for immersion freezing
+  real(r8),parameter :: theta_imm_bc = 48.0_r8      ! contact angle [deg], converted to rad later !DeMott et al (1990)
```
Why represent these parameters in degrees when they are only used in radians?
```suggestion
    real(r8),parameter :: deg2rad = pi / 180.0_r8 
    real(r8),parameter :: theta_imm_bc = 48.0_r8 * deg2rad      ! contact angle [rad] !DeMott et al (1990)
```
This suggestion applies to the other three constants immediately below.

mvertens replied:
This will change answers - so I'm very hesitant to do this unless you really feel its warranted. I'd like to keep this PR bfb.

mvertens replied:
Raised an issue for this.

gold2718 commented on src/oslo_aero_hetfrz.F90:
```diff
@@ -641,7 +743,7 @@ subroutine hetfrz_classnuc_oslo_save_cbaero(state, pbuf)
 
     ! local variables
-    integer :: i, lchnk, kk, ncol, m, n
+    integer :: i, k, lchnk, kk, ncol, m, n
```
Please do not introduce new, single letter variables.
```suggestion
    integer :: i, lev, lchnk, kk, ncol, m, n
```

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_hetfrz.F90:
```diff
@@ -654,201 +756,194 @@ subroutine hetfrz_classnuc_oslo_save_cbaero(state, pbuf)
-    !*****************************************************************************
-    !                calculate cloud borne aerosol
-    !*****************************************************************************
+    integer  :: i,k,m
```
Please replace single-letter variables with something indicating its use.

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_hetfrz.F90:
```diff
@@ -654,201 +756,194 @@ subroutine hetfrz_classnuc_oslo_save_cbaero(state, pbuf)
+          tmp2 = max(6.0_r8*dr_so4_monolayers_dust*volumeCore(i,k,num_bc_idx), 0.0_r8)
+          dstcoat(i,k,1) = tmp1/tmp2
+          if (dstcoat(i,k,1) > 1._r8) dstcoat(i,k,1) = 1._r8
+          if (dstcoat(i,k,1) < 0.001_r8) dstcoat(i,k,1) = 0.001_r8
```
Is this more efficient than a combined statement?
```suggestion
          dstcoat(i,k,1) = max(0.001_r8, min(1._r8, dstcoat(i,k,1)))
```
similarly for 2 and 3 below.

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_hetfrz.F90:
```diff
@@ -336,6 +381,67 @@ subroutine hetfrz_classnuc_oslo_init(mincld_in)
+    ! only consider flat surfaces due to uncertainty of curved surfaces
+
+    m = COS(theta_imm_bc*pi/180._r8)
+    f_imm_bc = (2+m)*(1-m)**2/4._r8
```
`m` is a real number. All of these equations should use real constants:
```suggestion
    f_imm_bc = (2._r8+m)*(1._r8-m)**2/4._r8
```

mvertens replied:
Changed m to cos_angle.

gold2718 commented on src/oslo_aero_linear_interp.F90:
```diff
@@ -1,6 +1,7 @@
 module  oslo_aero_linear_interp
 
   use shr_kind_mod, only: r8 => shr_kind_r8
+  use perf_mod,     only: t_startf, t_stopf
```
Not used, remove?

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_ndrop.F90:
```diff
@@ -93,16 +97,15 @@ module oslo_aero_ndrop
   subroutine ndrop_init_oslo()
 
-    integer            :: ii, l, lptr, m, mm
+    integer            :: ii, l, lptr, m, mm, k
```
Please do not introduce single-letter variables. How about `lind` or `lev`?

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_ndrop.F90:
```diff
@@ -357,22 +377,22 @@ subroutine dropmixnuc_oslo( state, ptend, dtmicro, pbuf, wsub,
-
+    real(r8) :: ekkp(pver)                      ! zn*zs*density*diffusivity (kg/m3 m2/s) at interface
+    real(r8) :: ekkm(pver)                      ! zn*zs*density*diffusivity (kg/m3 m2/s) at interface
+                                                ! above layer k  (k,k+1 interface)
```
I do not understand this comment. Does 'k+1 interface' refer to what the index would be for an interface variable? Something else?

gold2718 commented on src/oslo_aero_ndrop.F90:
```diff
@@ -513,37 +533,33 @@ subroutine dropmixnuc_oslo( state, ptend, dtmicro, pbuf, wsub,
-          tracerIndex      =  getTracerIndex(m,l,.false.)                   !Index in q
-          cloudTracerIndex =  getCloudTracerIndex(m,l)              !Index in phys-buffer
+          tracerIndex      =  tracer_index(m,l)                     !Index in q
+          cloud_tracer_index = cloudTracerIndex(tracerIndex)        !Index in phys-buffer
```
Please use consistent capitalization:
```suggestion
          cloud_tracer_index = CloudTracerIndex(tracerIndex)        !Index in phys-buffer
```
On the other hand, what is the value of making a copy of the variable (`CloudTracerIndex` is an integer array in oslo_aero_share.F90).

gold2718 commented on src/oslo_aero_ndrop.F90:
```diff
@@ -561,13 +577,13 @@ subroutine dropmixnuc_oslo( state, ptend, dtmicro, pbuf, wsub,
        do m=1,nmodes
-          if(.NOT. alert .and. &
-               ANY(numberConcentration(:ncol,k,m) .lt. 0.0_r8 ))then
+          if(.NOT. alert .and. ANY(numberConcentration(:ncol,k,m) .lt. 0.0_r8 ))then
```
Good time to switch to modern notation?
```suggestion
          if(.NOT. alert .and. ANY(numberConcentration(:ncol,k,m) < 0.0_r8 ))then
```

mvertens replied:
Done.

gold2718 commented on src/oslo_aero_ndrop.F90:
```diff
@@ -632,55 +653,98 @@ subroutine dropmixnuc_oslo( state, ptend, dtmicro, pbuf, wsub,
+                  componentFractionOK(m,tracerIndex,k) = &
+                       Cam(i,k,m) * f_acm(i,k,m) * (1.0_r8 -f_bcm(i,k,m)) * (f_soam(i,k,m)) &
+                       /(CProcessModes(i,k)*f_c(i,k)*(1.0_r8 -f_bc(i,k))*f_soa(i,k) + smallNumber)
+               end if
```
No `else` statement? If falling through is not an error, a comment to that effect would be helpful. Otherwise, please trap an error.

mvertens replied:
Raised an issue for this.

gold2718 commented on src/oslo_aero_ndrop.F90:
```diff
@@ -632,55 +653,98 @@ subroutine dropmixnuc_oslo( state, ptend, dtmicro, pbuf, wsub,
+               !calculate fraction of component "l" in mode "m" based on concentrations in clear air
+               tracerIndex = tracer_index(m,l)
+
+               if((l_so4_a1 == tracerIndex))then !so4 condensation
```
I think this sequence would look better as a `case select` statement.

Also, I think the format of these equations was more clear in the old version (from oslo_aero_share).

mvertens replied:
Cannot do case since l_so4_a2 etc are not constants.

mvertens replied:
I mean parameters instead of constants set at run time.

gold2718 commented on src/oslo_aero_ndrop.F90:
You can remove the `if` statement:
```suggestion
             do k = top_lev, pver-1
                   ! update droplet source - activation source in layer k involves particles from k+1
                   srcn(k) = srcn(k) + nact(k,m)*(raercol(k+1,mm,nsav))
             end do
             ! new formulation for k=pver
             tmpa = raercol(pver,mm,nsav)*nact(pver,m) + raercol_cw(pver,mm,nsav)*(nact(pver,m)-taumix_internal_pver_inv)
             srcn(pver) = srcn(pver) + max(0.0_r8, tmpa)
```

mvertens replied:
This no longer applies.

gold2718 commented on src/oslo_aero_ndrop.F90:
I think this equation is missing a term.

gold2718 commented on src/oslo_aero_ndrop.F90:
This equation is incomplete.

gold2718 commented on src/oslo_aero_ndrop.F90:
Change this loop to `pver-1` and keep the `pver` special case outside the loop.

gold2718 commented on src/oslo_aero_ndrop.F90:
Change this loop to `pver-1` and keep the `pver` special case outside the loop.

gold2718 commented on src_cam/radsw.F90:
```diff
@@ -256,7 +256,7 @@ subroutine rad_rrtmg_sw(lchnk,ncol       ,rrtmg_levs   ,r_state      , &
    real(r8) :: tauxar(pcols,0:pver) ! aerosol extinction optical depth
    real(r8) :: wa(pcols,0:pver) ! aerosol single scattering albedo
-   real(r8) :: ga(pcols,0:pver) ! aerosol assymetry parameter
+   real(r8) :: ga(pcols,0:pver) ! aerosol asymetry parameter
```
Incorrect spelling correction:
```suggestion
   real(r8) :: ga(pcols,0:pver) ! aerosol asymmetry parameter
```

mvertens replied:
Done.

gold2718 commented on src_cam/radiation.F90:
```diff
@@ -1265,6 +1265,13 @@ subroutine radiation_tend( &
               volc_ext_sun, volc_omega_sun, volc_g_sun, volc_ext_earth, volc_omega_earth, &
               aodvis, absvis)
 
+         !TODO (mvertens): should the following be added here?
+         rd%cld_tau_cloudsim(:ncol,:) = cld_tau(rrtmg_sw_cloudsim_band,:ncol,:)
+         rd%aer_tau550(:ncol,:)       = aer_tau(:ncol,:,idx_sw_diag)
+         rd%aer_tau400(:ncol,:)       = aer_tau(:ncol,:,idx_sw_diag+1)
+         rd%aer_tau700(:ncol,:)       = aer_tau(:ncol,:,idx_sw_diag-1)
+         ! OSLO_AERO end
```
Where did this come from?

mvertens replied:
@gold2718 - It came from updating the shadow file to the new cam version. I still am not sure we want it.

mvertens replied:
I think it should be there at this point.

gold2718 commented on src_cam/radiation.F90:
```diff
@@ -1319,17 +1327,18 @@ subroutine radiation_tend( &
+                  cam_out%solld, fns, fcns, Nday, Nnite,                     &
+                  IdxDay, IdxNite, su, sd, E_cld_tau=c_cld_tau,              &
+                  E_cld_tau_w=c_cld_tau_w, E_cld_tau_w_g=c_cld_tau_w_g,      &
+                  E_cld_tau_w_f=c_cld_tau_w_f, old_convert=.false.)
```
I would leave in the `idrf=.false.` to make it explicit.

mvertens replied:
Done.

gold2718 commented on src_cam/radiation.F90:
```diff
@@ -1344,45 +1353,8 @@ subroutine radiation_tend( &
-      call outfld('CABS550 ',clearabs550    ,pcols,lchnk)
-      call outfld('CABS550A',clearabs550alt ,pcols,lchnk)
-#endif
+      end if
```
Do we need to restore the `cloudfree`, `clearodvis`, `clearabsvis` etc. diagnostics?

mvertens replied:
See #33

gold2718 commented on src_cam/radiation.F90:
```diff
@@ -1398,29 +1370,19 @@ subroutine radiation_tend( &
-
-               call outfld('FLNT_DRF',flnt(:)    , pcols, lchnk)
-               call outfld('FLNTCDRF',rd%flntc(:), pcols, lchnk)
+               ! Note that aer_lw_abs is calculated in the call to oslo_aero_optical_params_calc
```
Does that replace the call to `rad_rrtmg_lw`? What about the `outfld` calls on lines 1411 and 1412?

mvertens replied:
@gold2718 - Can you please clarify what you mean by this?

gold2718 commented on src_cam/vertical_diffusion.F90:
```diff
@@ -1441,7 +1540,7 @@ subroutine vertical_diffusion_tend( &
   call outfld( 'DVV'          , ptend%v,                   pcols, lchnk )
   do m = 1, pcnst
-     call outfld( vdiffnam(m) , ptend%q(:,:,m),            pcols, lchnk )
+     call outfld( vdiffnam(m) , ptend%q(1,1,m),            pcols, lchnk )
```
Please fix the field indices:
```suggestion
     call outfld( vdiffnam(m) , ptend%q(:,:,m),            pcols, lchnk )
```

mvertens replied:
Done.

gold2718 reviewed (changes requested):
I have some items I think need to be changed, some suggestions, and a couple of questions.
This is a review of 317e4ff50cd2525955dfcd29400b108937ba163c

mvertens committed:
fixed problem in oslo_aero_ndrop - although this does not seem to have changed answers

mvertens committed:
addressed comments in PR

mvertens committed:
addressed issues in PR #31

mvertens committed:
addressed more issues in PR #31

mvertens committed:
replaced .gt. with >

mvertens committed:
replaced .gt. with > etc.

mvertens committed:
replaced .gt. with > etc.

mvertens committed:
replaced .gt. with > etc.

mvertens committed:
addressed more issues in PR#31

mvertens committed:
output FSDMS, etc on output history

Commit:
Correct some typos from refactor
Removed a few unused variables
Converted integers to correct real values (e.g., 1 ==> 1._r8)

gold2718 reviewed (approved):
Epic work, looks good now.
